src/iterative_ensemble_smoother/_iterative_ensemble_smoother.py

```diff
@@ -8,16 +8,28 @@
 
 rng = np.random.default_rng()
 
-from ._ies import InversionType, make_D, make_E, update_A
+from ._ies import InversionType, make_D, make_E, create_coefficient_matrix
 from iterative_ensemble_smoother.utils import (
-    _compute_AA_projection,
     _validate_inputs,
     _create_errors,
 )
 
 
-class IterativeEnsembleSmoother:
-    """IterativeEnsembleSmoother performs the update step of the iterative ensemble ensemble
+def _response_projection(
+    param_ensemble: npt.NDArray[np.double],
+) -> npt.NDArray[np.double]:
+    """A^+A projection is necessary when the parameter matrix has fewer rows than
+    columns, and when the forward model is non-linear. Section 2.4.3
+    """
+    _, _, vh = np.linalg.svd(
+        param_ensemble - param_ensemble.mean(axis=1, keepdims=True), full_matrices=False
+    )
+    projection: npt.NDArray[np.double] = vh.T @ vh
+    return projection
```

Comment:
Is this faster than just simply writing
```suggestion
    return np.linalg.pinv(param_ensemble) @ param_ensemble
```
And also is it correct, referring to Equation 37 of Evensen 2019? Seems to be missing the block identity-zero matrix?

Reply:
My understanding of this is that since the number of rows (m) is less than the number of columns (N), the last n - m columns of V will be multiplied by zero. 
We are here calling numpy svd with `full_matrices=False` which should do the same as using the block matrix.

Here's a proof by Python :)

```python
import numpy as np
rng = np.random.default_rng()

n = 10
N = 20
X = rng.normal(size=(n, N))

A = (X - X.mean(axis=1, keepdims=True)) / np.sqrt(N - 1)

# Numpy SVD using full_matrices=False
_, _, vh_false = np.linalg.svd(A, full_matrices=False)
assert vh_false.shape == (n, N)
proj_false = vh_false.T @ vh_false

# Numpy SVD using full_matrices=True
_, _, vh_true = np.linalg.svd(A, full_matrices=True)
assert vh_true.shape == (N, N)
proj_true = vh_true.T @ vh_true

# Projection using block matrix as in evensen2019
I_block = np.zeros(shape=(N, N))
I_block[:n, :][:, :n] = np.identity(n)
proj_block = vh_true.T @ I_block @ vh_true
assert proj_block.shape == (N, N)

# Making sure projection using block matrix is the same as using numpy SVD with full_matrices=False
assert np.isclose(proj_false, proj_block).all()

# pinv gives the same result as svd approach.
# There is already a test on main that verifies this,
# but I have chosen to keep the code as is for now.
proj_pinv = np.linalg.pinv(A) @ A
assert np.isclose(proj_pinv, proj_false).all()
```

Reply:
I've force pushed a fix where _response_projection uses np.linalg.pinv instead.
Also, I'm now dividing by sqrt(N-1) when calculating projection, but this does not really matter it seems:
https://github.com/equinor/iterative_ensemble_smoother/issues/74